Migrate timelock scheduling from blocks to drand rounds

carloskiron opened the pull request:
## Summary
This PR migrates the timelock transaction scheduling system from block-based to drand round-based approach, implementing proper drand integration with real-time round tracking.

## Key Changes

### 🔄 Backend Migration
- **Updated timelock dependency** to `@ideallabs/timelock.js@0.2.0-dev` with drand support
- **Added DrandService** for interfacing with drand Quicknet API (3-second rounds)
- **Migrated ExplorerService** to encrypt transactions using drand rounds instead of blocks
- **Updated domain models** to use rounds as primary scheduling mechanism
- **Removed feature flag** - timelock scheduling is now always available

### 🎨 Frontend Updates  
- **Schedule form** now asks for deadline round instead of block number
- **Real-time drand round display** in Network Activity Hub with live updates
- **Fixed React hydration error** by properly structuring layout components
- **Added current round indicator** with (Current: X,XXX) display

### ⚙️ Configuration
- **Environment variables** for drand API URL and Quicknet chain hash
- **Created `.env.local.example`** with all configuration options
- **Updated documentation** (README.md) with new features
- **Configurable drand network** support for different environments

### 🛠️ Developer Experience
- **Comprehensive debugging** for missing timelock pallets on chain
- **Error handling** with detailed console output for troubleshooting
- **TypeScript interfaces** for drand API responses
- **Clean dependency injection** setup for drand services

## New Features
- ⏰ **Real-time drand round monitoring** in activity hub (updates every 3 seconds)
- 🔐 **Round-based timelock encryption** using BLS12-381 curve with DrandIdentityBuilder
- 🌐 **Environment configuration** for different drand networks (testnet/mainnet)
- 🐛 **Debug output** showing available pallets when timelock pallet is missing

## Testing
- ✅ Build passes successfully
- ✅ UI displays current drand round in real-time
- ✅ Schedule form validates future rounds correctly
- ⚠️ Timelock pallet detection needs chain verification

## Notes
The implementation includes debugging output to identify the correct timelock pallet and method names on the Ideal Network chain. This may need adjustment once the actual pallet structure is confirmed.

Commit:
Migrate timelock scheduling from blocks to drand rounds

## Summary
- Migrated timelock transaction scheduling from block-based to drand round-based approach
- Updated to use @ideallabs/timelock.js 0.2.0-dev with drand support
- Added real-time drand round tracking in the UI
- Made timelock scheduling always available (removed feature flag)

## Key Changes
### Backend
- Updated ExplorerService to use drand rounds instead of blocks for timelock encryption
- Added DrandService for interfacing with drand Quicknet API
- Updated DelayedTransactionDetails domain model to use rounds
- Added proper error handling and debugging for missing timelock pallets

### Frontend
- Updated schedule form to ask for deadline round instead of block
- Added live drand round display in Network Activity Hub
- Shows current round with 3-second updates matching Quicknet frequency
- Fixed React hydration error by properly structuring layout components

### Configuration
- Added configurable environment variables for drand API and chain hash
- Created .env.local.example with all configuration options
- Updated documentation (README.md, CLAUDE.md) with new features
- Removed unnecessary FEATURE_SCHEDULE_TRANSACTION flag

## New Features
- Real-time drand round monitoring in activity hub
- Round-based timelock encryption using BLS12-381 curve
- Environment variable configuration for different drand networks
- Comprehensive debugging output for missing chain pallets

Commit:
Merge main into cm/drand-timelock-migration

Resolved conflicts:
- client-layout.tsx: Kept our approach without Inter font className (moved to root layout)
- network-activity/page.tsx: Combined drand service with new refresh trigger
- di-provider.tsx: Kept our explicit div wrapper approach for hydration consistency

Commit:
Update package-lock.json after merge and build

driemworks commented on src/services/ExplorerService.ts:
```diff
@@ -76,39 +84,107 @@ export class ExplorerService implements IExplorerService {
+      const callData = innerCall.toU8a();
+
+      // Generate a random key for encryption (32 bytes)
+      const key = crypto.getRandomValues(new Uint8Array(32));
```
The secret key should be computed using an hkdf, like so:

```
const computeSecretKey = async (seed) => {
    const hash = 'SHA-256'
    const length = 32
    const esk = await hkdf.compute(seed, hash, length, '')
    const key = Array.from(esk.key)
        .map((byte) => byte.toString(16).padStart(2, '0'))
        .join('')
    return key;
}
```

carloskiron replied:
change made!

Commit:
Implement user-friendly date/time picker for timelock scheduling

- Add react-datepicker dependency with custom styling for light/dark modes
- Create ScheduleTimeInput component with toggle between datetime and round modes
- Add comprehensive date/round conversion utilities with timezone support
- Update schedule page to use intuitive date/time interface instead of raw rounds
- Implement HKDF-based deterministic key derivation for transaction recovery

juangirini reviewed (approved):
LGTM

Commit:
Update timelock scheduling boundaries based on team feedback

- Reduce minimum scheduling time from 5 minutes to 2 minutes
- Remove maximum 30-day scheduling limitation (no upper bound)
- Update validation messages to reflect new constraints
- Adjust default values to match new minimum requirement

carloskiron commented on src/components/timelock/ScheduleTimeInput.tsx:
```diff
@@ -0,0 +1,290 @@
+
+      {/* Help Text */}
+      <div className="text-xs text-zinc-500">
+        <p>• Minimum: 5 minutes in the future • Maximum: 30 days in the future</p>
```
@driemworks changed to be without upper limit, so now can be scheduled to whatever time/round in the future.

Commit:
Made min schedule time a constant in drandTimeUtils

carloskiron commented on src/app/timelock/schedule/page.tsx:
```diff
@@ -24,31 +24,57 @@ import { Button } from '@/components/button';
+        setCurrentRound(current);
+        // Set initial round if not set
+        if (round === 0) {
+          setRound(current + 100); // Default to 5 minutes from now
```
@driemworks I changed to be a constant within drandTimeUtils

carloskiron committed:
formatting

carloskiron committed:
Fix help text to reflect new scheduling boundaries

- Update help text to show 2 minutes minimum instead of 5 minutes
- Update help text to show no maximum limit instead of 30 days
- Use constants for consistency in messaging
